core/src/main/java/com/linecorp/armeria/server/RoutingContext.java

```diff
@@ -81,6 +84,23 @@
      */
     List<MediaType> acceptTypes();
 
+    /**
+     * Returns a list of {@link LanguageRange}s that are specified in {@link HttpHeaderNames#ACCEPT_LANGUAGE}
+     * in the order of client-side preferences. If the client does not send the header, this will contain only a
+     * wild card {@link LanguageRange}.
+     */
+    List<LanguageRange> acceptLanguages();
+
+    /**
+     * This matches the {@link Locale}s supported by the server to
+     * the @{link HttpHeaderNames#ACCEPT_LANGUAGE} and returning the best match
+     * according to client preference.
+     * @param supportedLocales a list of locales supported by the server.
+     * @return The best matching locale or null if no locale matches.
```

Comment:
I'm curious if we should return `Locale.ROOT` instead of `null`. What do you think?

Reply:
If the nullability is the issue, I'd prefer `Optional<Locale>`, because I don't really see what ROOT would represent, except that nothing was found. It can't really be used to select what texts to show to the user. My completely made up statistics says that in 99.9% of the cases when this fails, the one calling the method will fall back to a specific Locale that they support and hope that the user can read it anyway.

I imagine the code looking something like
```kotlin
selectLocale(acceptedLanguages, supportedLanguages)
    ?: Locale.US
```
```java
Optional.ofNullable(selectLocale(acceptedLanguages, supportedLanguages))
    .orElse(Locale.US)
```

if the method returns Locale.ROOT, the check would be the (imho) less clear:
```java
Optional.of(selectLocale(acceptedLanguages, supportedLanguages))
    .filter((locale) -> !locale.equals(Locale.ROOT))
    .orElse(Locale.US)
```

Reply:
I see. Returning `null` sounds better.